### PR TITLE
added deployment configuration to ecs_service

### LINKED ISSUE #2434: ecs_service module - Add support for deployment configuration
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

`ecs_service` module
##### SUMMARY

The module does not support that a deployment configuration is provided. In particular, it does not allow to set the values for the minimum healthy percent and maximum percent.
##### DETAILS

For more information on this configuration, search for `--deployment-configuration` on [this AWS CLI documentation page](http://docs.aws.amazon.com/cli/latest/reference/ecs/create-service.html) or check the API reference for the explanation of the [DeploymentConfiguration](http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DeploymentConfiguration.html) data structure.
##### EXPECTED RESULTS

Being able to create/update an ECS service with the following task:

```
- name: Configure ECS service for the web application
  ecs_service:
    name: test-service
    cluster: test-cluster
    task_definition: test-task-definition
    desired_count: 3
    deployment_configuration:
      minimum_healthy_percent: 75
      maximum_percent: 150
    state: present
```
##### ACTUAL RESULTS

The module currently assumes the default values for the deployment configuration (`minimumHealthyPercent = 100` and `maximumPercent = 200`).

### PR DESCRIPTION
**Pull request by simplesteph**
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

ecs_service
##### ANSIBLE VERSION

```
ansible 2.3.0
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

Added deployment_configuration as an optional configuration parameter for the ecs_service module, which allows to set minimumHealthyPercent and maximumPercent in both create and update functions. It doesn't do anything beyond using the existing parameters for boto3. It checks for existence of parameters and raises an exception is the arguments are not named correctly

this fixes #2434

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @simplesteph. To the current maintainers, @Java1Guy please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by adrianmoisey** [comment]
:+1:

**Comment by zacblazic** [comment]
👍

**Comment by zacblazic** [comment]
Can I request that we correctly pass through the min & max parameters as integers? If we don't do this we cannot use variables to set the deployment configuration.

Expanding on your code, this is what I did for now:

```
def parse_deployment_configuration(self, config):
  c = {}
  if config:
    c['minimumHealthyPercent'] = int(config['minimum_healthy_percent'])
    c['maximumPercent'] = int(config['maximum_percent'])
  return c
```

I'm sure it can be improved though.

At the same time we may as well stick to Ansible's underscore naming convention:

```
deployment_configuration:
  minimum_healthy_percent: "{{ deployment.minimum_healthy_percent }}"
  maximum_percent: "{{ deployment.maximum_percent }}"
```

**Comment by simplesteph** [comment]
I first looked at using underscore but then ansible would say the changes "failed" in aws because the returned document by boto has camel case and not underscore. If you know a way to modify the code and make things have underscore, or can point me to another module that's doing that, that'll be great.

Regarding passing in variables, I'll have a try next week and fix the issue with int

**Comment by zacblazic** [comment]
Yeah, boto unfortunately requires camel case!  Please take a look at what I've done [here](https://github.com/ansible/ansible-modules-extras/compare/devel...zacblazic:ecs-service-params) with regards to type casting the deployment configuration and load balancer definitions. It also expects underscores with spaces and later converts everything to camel case.

Feel free to incorporate it here, or I can create a PR using your changes once they're merged.

**Comment by simplesteph** [comment]
Sorry I haven't had the chance to look at it just yet, I'm quite busy -_- . If you could do a PR that'll be great...

**Comment by simplesteph** [comment]
@zacblazic I checked your code, I think it doesn't allow for load_balancers param to be optional now. Can you try not specifying it on your end?

**Comment by simplesteph** [comment]
Thanks @zacblazic for your edits

**Comment by zacblazic** [comment]
Thanks for adding the changes in @simplesteph. 

Optional `load_balancers` are possible, these operations should return `None` if the original input `module.params['load_balancers']` is `None`:

```
load_balancers = type_mapper.map_complex_type(module.params['load_balancers'], LOAD_BALANCER_TYPE_MAP)
load_balancers = type_mapper.camelize(load_balancers)
```

I then later check and set them to an empty array if they are indeed `None`:

```
if load_balancers is None:
  load_balancers = []
```

**Comment by simplesteph** [comment]
Thanks!
What are the next steps to getting this merged in? New to the ansible
community

File: cloud/amazon/ecs_service.py [hunk 1]
```diff
@@ -346,35 +441,41 @@ def main():
             if not module.check_mode:
                 if module.params['load_balancers'] is None:
                     loadBalancers = []
```
Comment (zacblazic) [inline cloud/amazon/ecs_service.py]: This isn't being used, it should be more like [this](https://github.com/ansible/ansible-modules-extras/compare/devel...zacblazic:ecs-service-params#diff-411a22dc031b2d8f51d4b69e307537e5R434).

**Comment by zacblazic** [comment]
@gregdek I'm not sure who the current maintainers are but can we get some eyes on this?

File: cloud/amazon/ecs_service.py (hunk 1 above)
Comment (simplesteph) [reply cloud/amazon/ecs_service.py]: I had fixed the typo but forgot to push. thanks for that, that was why I was asking about empty load balancers

**Comment by gregdek** [comment]
@rmorlok ping, can you review?

**Comment by rmorlok** [comment]
@gregdek I've got an existing PR that that's been ready to ship for a while (https://github.com/ansible/ansible-modules-extras/pull/2356). Can we get that one merged to devel now that 2.2 has shipped so this can be rebased against that and then I take a look?

**Comment by zacblazic** [comment]
Small fix from my end that should be included: https://github.com/simplesteph/ansible-modules-extras/commit/180c4046b1ea7be23add4d69fffddd5eceac31a5

**Comment by gregdek** [comment]
@simplesteph the maintainer of this module has decided to step down and has recommended you as a potential maintainer. Would you be interested in that responsibility? More info here: 

https://github.com/ansible/ansible-modules-extras/blob/devel/GUIDELINES.md

Ping me if you'd be interested in helping out.

**Comment by simplesteph** [comment]
@gregdek Thanks for the opportunity to become a maintainer and sorry for the response delay. I accept and I've read the guidelines. Let me know what the next steps are. I'm monitoring ecs_service and ecs_taskdefinition mainly. Also, suggesting that @zacblazic helps if he's interested?

**Comment by kkarimi** [comment]
@simplesteph do you mind moving your PR to the [main ansible](https://github.com/ansible/ansible) repository? the modules repos have been merged there and PRs get merged much faster there.

**Pull request closed without merging**